bundle-workflow/src/test_workflow/test_recorder/test_recorder.py

```diff
@@ -0,0 +1,123 @@
+# SPDX-License-Identifier: Apache-2.0
+#
+# The OpenSearch Contributors require contributions made to
+# this file be licensed under the Apache-2.0 license or a
+# compatible open source license.
+
+import logging
+import os
+import shutil
+from typing import Dict
+
+import yaml
+
+from test_workflow.test_recorder.test_recorder_builder import \
+    TestRecorderBuilder
+
+
+class Singleton(type):
```

Comment:
+1

Reply:
Because we only need 1 instance of this class for each test_suite. Check the conversation here: https://github.com/opensearch-project/opensearch-build/pull/467#discussion_r708703896 for more info

Reply:
That's not a good enough reason to IMO. My concern is global objects that are accessed from various parts of the executing code, which makes it more difficult to test and reason about.

Reply:
So the reason behind it was as @gaiksaya said, we just want to have one instance of TestRecorder and make sure all the tests are recorded in one place. 
- This helps in putting a mechanism of just using one instance.
- Simpler for test publisher to sweep through the data and publish.

Reply:
https://stackoverflow.com/q/228164

TestPublisher should accept a 'ready' test recorder, otherwise how can it know if it published correctly.  What happens if a call is made to a test recorder after it was published, the lifecycle gets very confusing and requires adding more logic into this class that is not needed if its just a normal instanced object

Reply:
@saratvemulapalli i am not seeing much benefit of keeping this singleton. It also seems to create problem with parallelizing. What do you think about https://github.com/opensearch-project/opensearch-build/pull/587/files#r715918155 ?

Reply:
I agree, if we have a use case of having multiple instances will help then using Singleton is not our way to go. 
The only worry I have is when we have multiple instances of test recorder, test publisher, somebody has to maintain the sync between them.

Reply:
I didn't follow the need to sync them. In my understanding, each instance of test recorder/publisher can be record and publish independently to S3 at component level (`<base_path>/<component-X>/logs/`). There is no shared state between the test recorder/publisher instances. (PS - extra details from this point onward) At jenkins level, the test job completes once all components have been tested and their logs published. This is followed by notifications step

Reply:
I agree. TestRecorder and publisher should be treated as an independent component. How the tests use them is upon them. I'll remove the singleton pattern then. 

I tested it, before I made this class Singleton and there was passing around of testrecorder object which is fine I feel. Its gives us the clarity of code and helps us in debugging. This will help us when we make test run in parallel. Each of them will have its own instance of test_recorder.